src/persist-client/src/internal/trace.rs

```diff
@@ -1470,49 +1431,33 @@ impl<T: Timestamp + Lattice> MergeState<T> {
     ///
     /// There is the additional option of input batches.
     fn complete(&mut self, log: &mut SpineLog<'_, T>) -> Option<SpineBatch<T>> {
-        match std::mem::replace(self, MergeState::Vacant) {
-            MergeState::Vacant => None,
-            MergeState::Single(batch) => Some(batch),
-            MergeState::Double(a, b, merge) => Some(merge.done(a, b, log)),
+        let mut this = mem::take(self);
+        if let Some(merge) = this.merge {
+            merge.done(this.batches, log)
+        } else {
+            assert!(
+                this.batches.len() <= 1,
+                "Multi-batch runs should have an in-progress merge!"
+            );
```

Comment:
This isn't what we'd want in the general case, since when N > 2 a batch could have more than one batch but not be merging yet. I think the "right" way to implement this (given existing usage) is as a merge of all batches in the layer, even when there is no already-active merge.